### PR TITLE
Feat: Add todo functionality

### PR DESCRIPTION
**Pull request by mkh-wtag**
Add todo in the todo list

### CODE REVIEW

File: components/CreateTodo.js [hunk 1]
```diff
@@ -0,0 +1,118 @@
+
+import Toast from "./Toast.js";
+
+const CreateTodo = () => {
```
Comment (hhu-wtag) [inline components/CreateTodo.js]:
```suggestion
const createTodo = () => {
```

File: components/CreateTodo.js [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+  const buttonCreate = document.querySelector('#buttonCreate');
+  const createTodo = document.querySelector('.create-todo');
```
Comment (hhu-wtag) [inline components/CreateTodo.js]: can we also target createTodo using an id to be consistant along with other dom targets?

File: components/CreateTodo.js [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+  const cancelTodoBtn = document.querySelector('#cancelTodo');
```
Comment (hhu-wtag) [inline components/CreateTodo.js]:
```suggestion
  const cancelTodoButton = document.querySelector('#cancelTodo');
```

File: components/CreateTodo.js [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+        hour = currentHour - 12;
+        amPm = 'PM';
+      } else {
+        hour = currentHour;
+      }
+
+      return {hour, amPm};
+    }
+
+    const { hour, amPm } = hourCalculation();
```
Comment (hhu-wtag) [inline components/CreateTodo.js]: Can we move this into an utilities file?

File: components/CreateTodo.js [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+
+  const emptyNotice = document.createElement('h2');
+
+  if(todoWrapper.innerHTML === '') {
+    emptyNotice.innerText = 'Todo list is empty';
+    todoWrapper.appendChild(emptyNotice)
+  }
+
+  const addTodo = () => {
+    const messageConfirmation = Toast('Please enter todo', 'error');
```
Comment (hhu-wtag) [inline components/CreateTodo.js]:
```suggestion
    const emptyTodoErrorNotice = Toast('Please enter todo', 'error');
```

File: components/CreateTodo.js [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+
+    if(!textArea.value) {
+      todoWrapper.appendChild(messageConfirmation);
+      setTimeout(() => {
+        messageConfirmation.remove();
+      }, 2000)
+    }
```
Comment (hhu-wtag) [inline components/CreateTodo.js]: If no textArea value is provided we have to return early
```suggestion
    if(!textArea.value) {
      todoWrapper.appendChild(messageConfirmation);
      setTimeout(() => {
        messageConfirmation.remove();
      }, 2000)
      
      return
    }
```

File: components/CreateTodo.js
Comment (hhu-wtag) [inline components/CreateTodo.js:1]: In general let's not directly manipulate DOM but maintain a data structure to store all the todo info.
e.g:
```js
let todos = []

const todo = createNewTodo("New Task1", {isDone: false, isEditing: false, createdAt: getCurrentTimeFormatted()})
todos.push(todo)

// Render the list
renderTodoList()
```

**Review by hhu-wtag — CHANGES_REQUESTED** [review]

File: components/CreateTodo.js
Comment (mkh-wtag) [reply components/CreateTodo.js:1]: Tried to fix.

File: index.js [hunk 7]
```diff
@@ -0,0 +1,4 @@
+import CreateTodo from "./components/CreateTodo.js";
+
+CreateTodo();
```
Comment (hhu-wtag) [inline index.js]: As this is the entry point of our app, let's name it more appropriately.

```suggestion
InitializeTodoApp();
```

File: components/CreateTodo.js
Comment (hhu-wtag) [inline components/CreateTodo.js:1]: Can we follow a declarative approach to the architecture of the file?

Initially we have one entry point, e.g: `index.js`
```js
// index.js
import initializeTodoApp from "entry.js"

initializeTodoApp()
```

Inside `entry.js` file we have our various modules working together

```js
// entry.js
import { inputFieldElement, submitButton, cancelButton } from "../domElements.js"
import { createTodoItem, renderTodoList } from "../domActions.js"
import { isUserInputValid } from "../utilities.js"

export function initializeTodoApp(){
// Global Data
let todos = []

submitButton.addEventListener("click", handleTodoSubmit)
cancelButton.addEventListener("click", handleCancel)

const handleTodoSubmit = () => {
    const content = inputFieldElement.textContent
    if(isUserInputValid(content) === false){
       return 
    }
    
    const todoItem = {id=Date.now(), value: content, isDone: false, isEditing: false, createdAt: '2023'}
    todos.push(todoItem)


    renderTodoList(todos)
}
}
```

Example of `domElements.js`
```js
// domElements.js
export const addTodoButton = document.querySelector("#add-todo-button")
export const listWrapper = document.querySelector("#todo-list-wrapper")
```

Example of `domActions.js`
```js
// domActions.js
import listWrapper from "../domActions.js"

function createButton(content, icon, handler, class, id){
const button = document.createElement("button")
button.classList.push(class)
button.innerHTML = ``
button.textContent = content
button.addEventListener("click", () => handler(id))
}

export const createTodoItem = (payload) => {
 const todoItem = document.createElement("li")
 todoItem.innerHTML = `<div>${payload.value}</div>`

 const deleteButton = createButton("delete", deleteIcon, handleTodoDelete, payload.id)
 // ...

return todoItem
}

function handleTodoDelete(id){
 todos = todos.filter({(id: todoId}) => todoId != id)
 renderTodoList()
}

export const renderTodoList = (todos) => {
   todos.forEach(todo => {
        listWrapper.appendChild(createTodoItem(todo))
   })
}
```

The idea of this architecture is that we delegate work to different modules and re-use functionalities when we need them.

We also need to think about the data structure of the each todo items.
A todo item is not only a text content. 
It has some state, such as, 
1. Is it in editing state
2. Is it done
3. When was it created

We can encapsulate all this logic inside an object and store them in the `todos` global array and render them whenever we need.

e.g:
```js
const todoItem = createTodo("Hello World") // {value: "Hello world", isEditing: false, isDone: false, createdAt: "Some time value"}
```

**Review by hhu-wtag — CHANGES_REQUESTED** [review]

File: index.js [hunk 8]
```diff
@@ -0,0 +1,7 @@
+// import CreateTodo from "./components/CreateTodo.js";
+
+// CreateTodo();
+
```
Comment (hhu-wtag) [inline index.js]: We can remove the commented out code. Thanks

File: index.html [hunk 9]
```diff
@@ -9,10 +9,62 @@
...
+
+            <button id="cancelTodo" class="button button-icon" type="button">
+              <img src="./public/icon-delete.svg" alt="delete" />
+            </button>
+          </div>
+        </div>
+      </div>
+    </form>
+
+    <!-- <div class="toast-wrapper">Hello</div> -->
```
Comment (hhu-wtag) [inline index.html]: ditto. 
ditto means, the similar suggestion have been given earlier. In this case, we can remove the commented out code.

File: entry.js [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+}
+
+export default initializeTodoApp;
+
+
+
+
+
+
+
```
Comment (hhu-wtag) [inline entry.js]: Extra whitespace can be removed. Please use prettier plugin to automatically format the code.

File: entry.js [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+  todoWrapper,
+  emptyNotice,
+} from "./components/domElements.js";
+
+import { 
+  createTodoItem,
+  renderTodoList,
+ } from "./components/domActions.js";
+
+import { openCreateTodo, cancelTodo } from "./components/domActions.js";
```
Comment (hhu-wtag) [inline entry.js]: They seem to import from the same file. We can use one statement.
```suggestion
import { 
  createTodoItem,
  renderTodoList,
  openCreateTodo,
  cancelTodo
 } from "./components/domActions.js";
```

File: components/domActions.js [hunk 12]
```diff
@@ -0,0 +1,85 @@
...
+};
+
+
+
+
+
+
+
+
+
```
Comment (hhu-wtag) [inline components/domActions.js]: ditto. Use prettier

File: components/CreateTodo.js [hunk 13]
```diff
@@ -0,0 +1,111 @@
...
+  // cancelTodoButton.addEventListener('click', cancelTodo);
+
+  // const emptyNotice = document.createElement('h2');
+
+  // if(todoWrapper.innerHTML === '') {
+  //   emptyNotice.innerText = 'Todo list is empty';
+  //   todoWrapper.appendChild(emptyNotice)
+  // }
+
+  let todos = [];
```
Comment (hhu-wtag) [inline components/CreateTodo.js]: We should not declare the source of data here. `todos` array should be initialised higher up the tree / as close to root as possible. 

So can we move it to `entry.js` and use it there?

```js
// entry.js file
import blah blah

// We declare the source of data
let todos = []

// We add the eventListeners
const addButton.addEventListener("click", handleAddTodo)

File: components/CreateTodo.js [hunk 14]
```diff
@@ -0,0 +1,111 @@
...
+    //   <button class="button button-icon">
+    //     <img src="./public/icon-edit.svg" alt="edit" />
+    //   </button>
+
+    //   <button class="button button-icon">
+    //     <img src="./public/icon-delete.svg" alt="delete" />
+    //   </button>
+    // `
+    
+    // const todoSuccess = toast('Successfully created new todo');
```
Comment (hhu-wtag) [inline components/CreateTodo.js]: Let's remove the unused comments

File: components/domActions.js [hunk 15]
```diff
@@ -0,0 +1,85 @@
...
+    createdAt: timeCalculation(),
+  }
+
+  todos.push(todo);
+
+  textArea.value = '';
+  emptyTodoError('Successfully created new todo');
+
+  return todos;
+};
```
Comment (hhu-wtag) [inline components/domActions.js]: `renderTodoList` is supposed to be adding actual DOM elements to the document.
Let's rename the function to `createTodoObject`
```suggestion
export const createTodoObject = () => {
  const task = textArea.value;

  const todo = {
    title: task,
    isDone: false,
    isEditing: false,
    createdAt: timeCalculation(),
  }
};
```

Notice that we wouldn't do anything else in this function following Single Responsibility principle.

File: components/domActions.js [hunk 16]
```diff
@@ -0,0 +1,85 @@
...
+  todoDetails.classList.add('todo-details');
+
+  const todoActions = document.createElement('div');
+  todoActions.classList.add('todo-actions');
+
+  const taskCompleted = createButton('button button-icon', 'taskCompleted', 'icon-tick.svg');
+  const taskEdit = createButton('button button-icon', 'taskEdit', 'icon-edit.svg' );
+  const taskDelete = createButton('button button-icon', 'taskDelete', 'icon-delete.svg' );
+
+  const todos = renderTodoList();
```
Comment (hhu-wtag) [inline components/domActions.js]:
```suggestion
```

File: components/domActions.js [hunk 17]
```diff
@@ -0,0 +1,85 @@
...
+      todoDetails.innerHTML = `<div class="created-at">Created at: <strong>${hour}:${minutes} ${amPm}, ${currentDay}-${months[currentMonth]}-${currentYear}</strong></div>`;
+    })
+  }
+  
+  todoActions.append(taskCompleted, taskEdit, taskDelete);
+  todoDiv.append(todoHeader, todoDetails, todoActions);
+  todoWrapper.append(todoDiv);
+
+  todoDiv.style.borderLeft = '5px solid #0ec277';
+  todoDiv.style.background = 'linear-gradient(90deg, #daf5ea, #fff 15%)';
```
Comment (hhu-wtag) [inline components/domActions.js]: We'll do the mapping inside the function named `renderTodoList`
So everything in one go:
1. We add something to the inputbox and click submit
2. It triggers the eventListener we added for submit button
3. That eventListener handler function first calls `createTodoObject`
4. We then push that todo object to todos array
5. Then we call `renderTodoList` which will
    i. Map over each todo object from `todos` array and create DOM element using `createTodoItem` function
    ii. Then insert it to DOM

**Review by hhu-wtag — CHANGES_REQUESTED** [review]

File: components/CreateTodo.js [hunk 18]
```diff
@@ -0,0 +1,117 @@
...
+
+  // buttonCreate.addEventListener('click', openCreateTodo);
+  // cancelTodoButton.addEventListener('click', cancelTodo);
+
+  // const emptyNotice = document.createElement('h2');
+
+  // if(todoWrapper.innerHTML === '') {
+  //   emptyNotice.innerText = 'Todo list is empty';
+  //   todoWrapper.appendChild(emptyNotice)
+  // }
```
Comment (hhu-wtag) [inline components/CreateTodo.js]: Can we remove the commented out code?

File: components/CreateTodo.js [hunk 19]
```diff
@@ -0,0 +1,117 @@
...
+    // const todoDetails = document.createElement('div');
+    // todoDetails.classList.add('todo-details');
+
+    // const todo = {
+    //   title: textArea.value,
+    //   isDone: false,
+    //   isEditing: false,
+    //   createdAt: timeCalculation(),
+    // }
+    // todos.push(todo);
```
Comment (hhu-wtag) [inline components/CreateTodo.js]: ditto

File: components/CreateTodo.js [hunk 20]
```diff
@@ -0,0 +1,117 @@
...
+
+    todos.map((todo) => {
```
Comment (hhu-wtag) [inline components/CreateTodo.js]:
```suggestion
    todos.forEach((todo) => {
```

File: components/CreateTodo.js [hunk 21]
```diff
@@ -0,0 +1,117 @@
...
+    //   <button class="button button-icon">
+    //     <img src="./public/icon-edit.svg" alt="edit" />
+    //   </button>
+
+    //   <button class="button button-icon">
+    //     <img src="./public/icon-delete.svg" alt="delete" />
+    //   </button>
+    // `
+
+    // const todoSuccess = toast('Successfully created new todo');
```
Comment (hhu-wtag) [inline components/CreateTodo.js]: ditto

File: components/CreateTodo.js [hunk 22]
```diff
@@ -0,0 +1,117 @@
...
+      todoDiv.style.borderLeft = "5px solid #0ec277";
+      todoDiv.style.background = "linear-gradient(90deg, #daf5ea, #fff 15%)";
+
+      textArea.value = "";
+      emptyNotice.remove();
+      setTimeout(() => {
+        todoSuccess.remove();
+        todoDiv.removeAttribute("style");
+      }, 2000);
+    }
```
Comment (hhu-wtag) [inline components/CreateTodo.js]: Let's use early return here

```suggestion
    if (!textArea.value) {
      return
    }
    
    todoDiv.append(todoHeader, todoDetails, todoActions);
    todoWrapper.append(todoDiv, todoSuccess);

    todoDiv.style.borderLeft = "5px solid #0ec277";
    todoDiv.style.background = "linear-gradient(90deg, #daf5ea, #fff 15%)";

     textArea.value = "";
     emptyNotice.remove();
     setTimeout(() => {
        todoSuccess.remove();
        todoDiv.removeAttribute("style");
      }, 2000);
```

File: components/domActions.js [hunk 23]
```diff
@@ -0,0 +1,67 @@
...
+
+export const renderTodoList = () => {
+  todoWrapper.innerHTML = "";
+
+  if (todos) {
+    todos.forEach((todo) => {
+      todoWrapper.append(createTodoDomElement(todo));
+    });
+  }
+};
```
Comment (hhu-wtag) [inline components/domActions.js]: We don't need the `if` check as we know that `todos` is always an array
```suggestion
export const renderTodoList = () => {
  todoWrapper.innerHTML = "";
  
  todos.forEach((todo) => {
      todoWrapper.append(createTodoDomElement(todo));
   });
};
```

File: components/CreateTodo.js
Comment (hhu-wtag) [inline components/CreateTodo.js:1]: Maybe we don't need this file anymore as we are currently creating todo dom element in `domActions.js`?

File: index.js (hunk 8 above)
Comment (hhu-wtag) [reply index.js]: Can we resolve this?

File: utilities/createButton.js [hunk 24]
```diff
@@ -0,0 +1,10 @@
+const createButton = (className, idName, icon, handler) => {
+  const button = document.createElement("button");
+  button.className = className;
+  button.id = idName;
+  button.innerHTML = `<img src=./public/${icon} alt="icon" />`;
+  button.addEventListener("click", handler);
+  return button;
+};
```
Comment (hhu-wtag) [inline utilities/createButton.js]:
```suggestion
const createButton = (className, idName, icon, handler) => {
  const button = document.createElement("button");
  button.className = className;
  button.id = idName;
  button.innerHTML = `<img src=./public/${icon} alt="icon" />`;
  button.addEventListener("click", handler);
  
  return button;
};
```

**Review by hhu-wtag — APPROVED** [review]
Please resolve the reviewes before merging the PR. Nice job bhai 🚀

**Pull request merged**